Logging/LogHandler/FileLogHandler.cs

```diff
@@ -94,9 +163,54 @@ protected override void DisposeSelf()
             string context = string.Format(
                 LogContextFormat,
                 callerDerivedTypeName, filename, callerName, callerLine
-                );
+            );
 
             m_Writer.WriteLine($"{timestamp}{logLevel}{context}{message}");
+
+            if (m_ShouldRotate && m_Writer.BaseStream.Length > RotateFileSizeLimit)
+            {
+                this.RotateFiles();
+            }
+        }
+
+        private void RotateFiles()
+        {
+            Debug.Assert(RotateFileSizeLimit >= 0, $"Expected positive integer for file size limit, but got {RotateFileSizeLimit}");
+            Debug.Assert(RotateFileCountLimit >= 0, $"Expected positive integer for file count limit, but got {RotateFileCountLimit}");
+
+            m_Writer.Close();
+            m_Writer = null;
+
+            // Determine how many rotated log files already exist
+            int index = 1;
+            string rotatedFilePath = GetRotatedFilePath(index);
+            while (File.Exists(rotatedFilePath))
+            {
+                rotatedFilePath = GetRotatedFilePath(++index);
+            }
+            --index;
+
+            // Delete any files in excess of the file count limit
+            while (index >= RotateFileCountLimit)
+            {
+                File.Delete(GetRotatedFilePath(index));
+                --index;
+            }
+
+            // Rotate remaining files
+            while (index >= 0)
+            {
+                File.Move(GetRotatedFilePath(index), GetRotatedFilePath(index + 1));
+                --index;
+            }
+
+            CreateWriter();
+        }
+
+        private string GetRotatedFilePath(int index)
+        {
+            Debug.Assert(index >= 0, $"Failed to get rotated file path, invalid index {index}");
+            return (index == 0 ? m_Path : Path.Combine(m_Directory, $"{m_FileName}.{index}{m_Extension}"));
         }
```

Comment:
For what it's worth, I tested paths with no directory or extension, i.e. if you give the path "logfile" this will still work, and your rotated files will be "logfile.1" etc.

Also if you give a path under several folders that don't exist yet, ex. "some/new/folders/log.txt", the `Directory.CreateDirectory(...)` in the constructor creates each folder in the path before opening the file writer (this was technically a bug in the previous code).

Reply:
Test cases are now summarized in PR description